RHIDP-2023: Added a procedure for running RHDH behing a corporate proxy (#225)
redhat-developer/red-hat-developers-documentation-rhdh: modules/admin/proc-running-rhdh-behind-a-proxy.adoc

[id='proc-running-rhdh-behind-a-proxy_{context}']
= Running the {product-very-short} application behind a corporate proxy

You can run the {product-very-short} application behind a corporate proxy by setting any of the following environment variables before starting the application:

* `HTTP_PROXY`: Denotes the proxy to use for HTTP requests.
* `HTTPS_PROXY`: Denotes the proxy to use for HTTPS requests.

Additionally, you can set the `NO_PROXY` environment variable to exclude certain domains from proxying. The variable value is a comma-separated list of hostnames that do not require a proxy to get reached, even if one is specified.

For Helm-based deployment, both a developer or a cluster administrator, with permissions to create resources in the cluster, can configure the proxy variables in a `values.yaml` Helm configuration file.

For Operator-based deployment, the approach you use for proxy configuration is based on your role:

> * As a cluster administrator with access to the Operator namespace, you can configure the proxy variables in the Operator's default ConfigMap file. This configuration applies the proxy settings to all the users within the namespace where the Operator is deployed.

Reviewer: > This configuration applies the proxy settings to all the users within the namespace where the Operator is deployed.

Just a small clarification on this as I think the "within the namespace ..." part can be misleading: I would say something like this: "This configuration applies the proxy settings to all the users of the operator.".

As the operator watches custom resources cluster-wide, the proxy settings configuration (once added by an admin to the operator default config) will be applied to all the users, regardless of the namespace in which they deploy their operator-based RHDH application (not necessarily within the namespace where the operator is deployed).

Author: Thanks, @rm3l, for the clarification and making me understand this part :)

> * As a developer, you can configure the proxy variables in a custom resource (CR) file. This configuration applies the proxy settings only to the users with access to that CR.

Reviewer: ```suggestion
* As a developer, you can configure the proxy variables in a custom resource (CR) file. This configuration applies the proxy settings to the {product-very-short} application created from that CR.
```

In the end, the proxy settings configured by the developer are applied to the RHDH application.

+
.Helm deployment

* Set the proxy information in your Helm configuration file:
+
[source,yaml]
----
upstream:
  backstage:
    extraEnvVars:
      - name: HTTP_PROXY
        value: '<http_proxy_url>'
      - name: HTTPS_PROXY
        value: '<https_proxy_url>'
      - name: NO_PROXY
        value: '<no_proxy_settings>'
----
+
Where,

`<http_proxy_url>`:: Denotes a variable that you must replace with the HTTP proxy URL.
`<https_proxy_url>`:: Denotes a variable that you must replace with the HTTPS proxy URL.
`<no_proxy_settings>`:: Denotes a variable that you must replace with comma-separated URLs, which you want to exclude from proxying, for example, `foo.com,baz.com`.
+
.Example: Setting proxy variables using Helm Chart

[source,yaml]
----
upstream:
  backstage:
    extraEnvVars:
      - name: HTTP_PROXY
        value: 'http://10.10.10.105:3128'
      - name: HTTPS_PROXY
        value: 'http://10.10.10.106:3128'
      - name: NO_PROXY
        value: 'localhost,example.org'
----

+
.Operator deployment 

** Perform one of the following steps based on your role:
* As an administrator, set the proxy information in the Operator's default ConfigMap file:
+

> .. Search for a ConfigMap file named `backstage-default-config` in the namespace where the Operator is deployed and open it. 

Reviewer: ```suggestion
.. Search for a ConfigMap file named `backstage-default-config` in the namespace where the Operator is deployed (`rhdh-operator` by default) and open it. 
```

The namespace where the RHDH operator is `rhdh-operator` by default. Should we add that here?

Author: Since we follow mod doc structure, we require to ensure that all the info we provide in a module is complete in itself. so, addint this info wud definitely be helpful.

.. Find the `deployment.yaml` key.
.. Set the value of the `HTTP_PROXY`, `HTTPS_PROXY`, and `NO_PROXY` environment variables in the `Deployment` spec as shown in the following example:
+ 
.Example: Setting proxy variables in a ConfigMap file
[source,yaml]
----
# Other fields omitted
  deployment.yaml: |-
    apiVersion: apps/v1
    kind: Deployment

>     # Other fields omitted

Reviewer: Even if the other fields are omitted, do you want to add the top-level fields as well, to make it clear what the path is? Something like this?
```yaml
# Other fields omitted
  deployment.yaml: |-
    apiVersion: apps/v1
    kind: Deployment
    spec:
      template:
        spec:
          initContainers:
            - name: install-dynamic-plugins
              env:
                - name: NPM_CONFIG_USERCONFIG
                  value: /opt/app-root/src/.npmrc.dynamic-plugins
                - name: HTTP_PROXY
                  value: 'http://10.10.10.105:3128'
                - name: HTTPS_PROXY
                  value: 'http://10.10.10.106:3128'
                - name: NO_PROXY
                  value: 'localhost,example.org'
          containers:
            - name: backstage-backend
              env:
                - name: APP_CONFIG_backend_listen_port
                  value: "7007"
                - name: HTTP_PROXY
                  value: 'http://10.10.10.105:3128'
                - name: HTTPS_PROXY
                  value: 'http://10.10.10.106:3128'
                - name: NO_PROXY
                  value: 'localhost,example.org'
```

              env:
                - name: NPM_CONFIG_USERCONFIG
                  value: /opt/app-root/src/.npmrc.dynamic-plugins
                - name: HTTP_PROXY
                  value: 'http://10.10.10.105:3128'
                - name: HTTPS_PROXY
                  value: 'http://10.10.10.106:3128'
                - name: NO_PROXY
                  value: 'localhost,example.org'

>   # Other fields omitted

Reviewer: There are 2 containers where the proxy settings need to be added. The one above is for the init container, but we also need to add them to the main application container.

[id='proc-running-rhdh-behind-a-proxy_{context}']
= Running the {product-very-short} application behind a corporate proxy

You can run the {product-very-short} application behind a corporate proxy by setting any of the following environment variables before starting the application:

* `HTTP_PROXY`: Denotes the proxy to use for HTTP requests.
* `HTTPS_PROXY`: Denotes the proxy to use for HTTPS requests.

Additionally, you can set the `NO_PROXY` environment variable to exclude certain domains from proxying. The variable value is a comma-separated list of hostnames that do not require a proxy to get reached, even if one is specified.

> For Helm-based deployment, both a developer or a cluster administrator, with permissions to create resources in the cluster, can configure the proxy variables in a `values.yaml` Helm configuration file.

Reviewer 2: ```suggestion
For Helm-based deployment, either a developer or a cluster administrator with permissions to create resources in the cluster can configure the proxy variables in a `values.yaml` Helm configuration file.
```

For Operator-based deployment, the approach you use for proxy configuration is based on your role:

* As a cluster administrator with access to the Operator namespace, you can configure the proxy variables in the Operator's default ConfigMap file. This configuration applies the proxy settings to all the users of the Operator.
* As a developer, you can configure the proxy variables in a custom resource (CR) file. This configuration applies the proxy settings to the {product-very-short} application created from that CR.

.Prerequisites

> * You have installed the {product} application. For more information, see xref:proc-install-rhdh_rhdh-getting-started[].

Reviewer 2: ```suggestion
* You have installed the {product} application. For more information, see xref:proc-install-rhdh_rhdh-getting-started[Administration guide for {product}].
```

Otherwise it shows up as _see [proc-install-rhdh_rhdh-getting-started]_ in the preview. 
Although I think you could also remove this xref entirely. It isn't really clear what the user is supposed to look at or gain from this guide.

.Procedure

. Configure proxy information in one of the following deployments based on your needs:
+
.Helm deployment

* Set the proxy information in your Helm configuration file:
+
[source,yaml]
----
upstream:
  backstage:
    extraEnvVars:
      - name: HTTP_PROXY
        value: '<http_proxy_url>'
      - name: HTTPS_PROXY
        value: '<https_proxy_url>'
      - name: NO_PROXY
        value: '<no_proxy_settings>'
----
+
Where,

`<http_proxy_url>`:: Denotes a variable that you must replace with the HTTP proxy URL.
`<https_proxy_url>`:: Denotes a variable that you must replace with the HTTPS proxy URL.
`<no_proxy_settings>`:: Denotes a variable that you must replace with comma-separated URLs, which you want to exclude from proxying, for example, `foo.com,baz.com`.
+
.Example: Setting proxy variables using Helm Chart

[source,yaml]
----
upstream:
  backstage:
    extraEnvVars:
      - name: HTTP_PROXY
        value: 'http://10.10.10.105:3128'
      - name: HTTPS_PROXY
        value: 'http://10.10.10.106:3128'
      - name: NO_PROXY
        value: 'localhost,example.org'
----

+

> .Operator deployment 

Reviewer 2: I recommend splitting the Helm and Operator procedures into two separate proc modules at some point. Perhaps we create a separate ticket for this so we don't block this merge.

** Perform one of the following steps based on your role:

> * As an administrator, set the proxy information in the Operator's default ConfigMap file:

Reviewer 2: I would consider creating separate proc modules for each of these role-based procedures also and specifying the role requirements in the Prerequisites section. However, that is an option we can address in the future.

[id='proc-running-rhdh-behind-a-proxy_{context}']
= Running the {product-very-short} application behind a corporate proxy

You can run the {product-very-short} application behind a corporate proxy by setting any of the following environment variables before starting the application:

* `HTTP_PROXY`: Denotes the proxy to use for HTTP requests.
* `HTTPS_PROXY`: Denotes the proxy to use for HTTPS requests.

Additionally, you can set the `NO_PROXY` environment variable to exclude certain domains from proxying. The variable value is a comma-separated list of hostnames that do not require a proxy to get reached, even if one is specified.

For Helm-based deployment, both a developer or a cluster administrator, with permissions to create resources in the cluster, can configure the proxy variables in a `values.yaml` Helm configuration file.

For Operator-based deployment, the approach you use for proxy configuration is based on your role:

* As a cluster administrator with access to the Operator namespace, you can configure the proxy variables in the Operator's default ConfigMap file. This configuration applies the proxy settings to all the users of the Operator.

> * As a developer, you can configure the proxy variables in a custom resource (CR) file. This configuration applies the proxy settings to the{product-very-short} application created from that CR.

Reviewer: ```suggestion
* As a developer, you can configure the proxy variables in a custom resource (CR) file. This configuration applies the proxy settings to the {product-very-short} application created from that CR.
```

.Prerequisites

* You have installed the {product}. For more information, see xref:proc-install-rhdh_rhdh-getting-started[].

> * You have created a project in the {product-short}.

Reviewer: Not sure what the second prerequisite means. We cannot create a project in Developer Hub, no?
